mitodl/ocw-studio
# automate transcript upload

#### Pre-Flight checklist


#### What are the relevant tickets?
https://github.com/mitodl/ocw-studio/issues/752

#### What's this PR do?
This pr adds code to automatically upload videos to 3play for transcription after they are created on youtube. 

#### How should this be manually tested?
- Set PREPUBLISH_ACTIONS=videos.tasks.update_transcripts_for_website,videos.youtube.update_youtube_metadata
Set THREEPLAY_API_KEY to the "api_key" value from 3play/project api key sandbox/3play_sandbox_api_key file in the team keybase
Set THREEPLAY_PROJECT_ID=51817
Set OCW_IMPORT_STARTER_SLUG=ocw-course

- Create your own youtube api configuration following the steps in the README section
- Get the short_id of a new or existing Website
- Go to https://drive.google.com/drive/folders/<DRIVE_SHARED_ID>/<short_id>/videos_final and add a video
- Wait a few minutes. Eventually there should be Video and VideoJob objects for that video.
- Check that the AWS mediaconvert job completed with
```
client = boto3.client(
     "mediaconvert",
     region_name=settings.AWS_REGION,
     aws_access_key_id=settings.AWS_ACCESS_KEY_ID,
     aws_secret_access_key=settings.AWS_SECRET_ACCESS_KEY,
     endpoint_url="https://vasjpylpa.mediaconvert.us-east-1.amazonaws.com"
 )
client.get_job(Id=VideoJob.objects.order_by("-created_on").first().job_id)
```
- Simulate the aws mediaconvert webhook by making a post to `https://localhost:8043/api/transcode-jobs`
with
```
{
 "version": "0",
 "id": "426c759f-277e-ee0f-aceb-7fb79214bc51",
 "detail-type": "MediaConvert Job State Change",
 "source": "aws.mediaconvert",
 "account": "<settings.AWS_ACCOUNT_ID>"
 ",
 "detail": {
 	"timestamp": 1629911639065,
 	"accountId": "919801701561",
 	"queue": "arn:aws:mediaconvert:us-east-1:919801701561:queues/Default",
 	"jobId": "<VideoJob.job_id>",
 	"status": "COMPLETE",
 	"userMetadata": {},
 	"outputGroupDetails": [{
 		"outputDetails": [{
 				"outputFilePaths": [
 					"s3://<settings.AWS_STORAGE_BUCKET_NAME>/aws_mediaconvert_transcodes/<Website.short_id>/<DriveFile.file_id>/<original_video_filename_base>_youtube.mp4"
 				],
 				"durationInMs": 45466,
 				"videoDetails": {
 					"widthInPx": 320,
 					"heightInPx": 176
 				}
 			},
 			{
 				"outputFilePaths": [
 					"s3://<settings.AWS_STORAGE_BUCKET_NAME>/aws_mediaconvert_transcodes/<Website.short_id>/<DriveFile.file_id>/<original_video_filename_base>_360p_16_9.mp4"
 				],
 				"durationInMs": 45466,
 				"videoDetails": {
 					"widthInPx": 640,
 					"heightInPx": 360
 				}
 			},
 			{
 				"outputFilePaths": [
 					"s3://<settings.AWS_STORAGE_BUCKET_NAME>/aws_mediaconvert_transcodes/<Website.short_id>/<DriveFile.file_id>/<original_video_filename_base>_360p_4_3.mp4"
 				],
 				"durationInMs": 45466,
 				"videoDetails": {
 					"widthInPx": 480,
 					"heightInPx": 360
 				}
 			}
 		],
 		"type": "FILE_GROUP"
 	}]
 }
}
```
as the payload.
- You should see `upload_youtube_videos`, `update_youtube_statuses ` and `start_transcript_job` tasks
- The status of the Video object created for your video should be "SUBMITTED_FOR_TRANSCRIPTION"
- You can check that there is a pending transcript by making a get request to`https://api.3playmedia.com/v3/transcripts/?media_file_reference_id=<the youtube id of your video>&api_key=<THREEPLAY_API_KEY>`
- Your youtube video will upload as private,  so the above transcription job will fail
- To test the 3play callback, set the `destination_id` for the youtube VideoFile for your video to `2gsx-zHwAWs`. Go to the site UI and set the Youtube ID for the video resource to `2gsx-zHwAWs` as well.
- Make a post request to `http://localhost:8043/api/transcription-jobs/?video_id=<the id of the Video object>`
- Go to the resource page for your video. The Video Captions and Video Transcript sections should be filled out
